internal/controller/dragonfly_controller.go

```diff
@@ -88,7 +89,128 @@ func (r *DragonflyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			return ctrl.Result{}, err
 		}
 
-		r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Created", "Created resources for Dragonfly object")
+		r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Resources", "Created resources")
+		return ctrl.Result{}, nil
+	} else if df.Status.IsRollingUpdate {
+		// This is a Rollout
+		log.Info("Rolling out new version")
+		var updatedStatefulset appsv1.StatefulSet
+		if err := r.Get(ctx, client.ObjectKey{Namespace: df.Namespace, Name: df.Name}, &updatedStatefulset); err != nil {
+			log.Error(err, "could not get statefulset")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		// get pods of the statefulset
+		var pods corev1.PodList
+		if err := r.List(ctx, &pods, client.InNamespace(df.Namespace), client.MatchingLabels(map[string]string{
+			"app":                               df.Name,
+			resources.KubernetesAppNameLabelKey: "dragonfly",
+		})); err != nil {
+			log.Error(err, "could not list pods")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		if len(pods.Items) != int(*updatedStatefulset.Spec.Replicas) {
+			log.Info("Waiting for all replicas to be ready")
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// filter replicas to master and replicas
+		var master corev1.Pod
+		replicas := make([]corev1.Pod, 0)
+		for _, pod := range pods.Items {
+			if _, ok := pod.Labels[resources.Role]; ok {
+				if pod.Labels[resources.Role] == resources.Replica {
+					replicas = append(replicas, pod)
+				} else if pod.Labels[resources.Role] == resources.Master {
+					master = pod
+				}
+			} else {
+				log.Info("found pod without label", "pod", pod.Name)
+				// retry after they are ready
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+			}
+		}
+
+		// We want to update the replicas first then the master
+		// We want to have at most one updated replica in full sync phase at a time
+		// if not, requeue
+		fullSyncedUpdatedReplicas := 0
+		for _, replica := range replicas {
+			// Check only with latest replicas
+			onLatestVersion, err := isPodOnLatestVersion(ctx, r.Client, &replica, &updatedStatefulset)
+			if err != nil {
+				log.Error(err, "could not check if pod is on latest version")
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, err
+			}
+			if onLatestVersion {
+				// check if the replica had a full sync
+				log.Info("New Replica found. Checking if replica had a full sync", "pod", replica.Name)
+				err := checkForStableState(ctx, r.Client, &replica)
```

Comment:
I'd rather we change the checkForStableState signature from returning error or nil to return bool, err so we can distinguish between false and error and also I think is clearer to the reader.

Reply:
Updated